cupy/core/core.pyx

```diff
@@ -1984,18 +1984,120 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
 
 
 cpdef ndarray concatenate(tup, axis, shape, dtype):
-    cdef ndarray ret
+    cdef ndarray a, x, ret
+    cdef int i, base, cum
+    cdef bint all_same_type, all_one_and_contiguous
+
     ret = ndarray(shape, dtype=dtype)
 
-    skip = (slice(None),) * axis
-    i = 0
+    all_same_type = True
+    all_one_and_contiguous = True
+    dtype = tup[0].dtype
     for a in tup:
-        aw = a.shape[axis]
-        ret[skip + (slice(i, i + aw),)] = a
-        i += aw
+        all_same_type &= a.dtype == dtype
```

Comment:
How about using `all_same_type = all_same_type or a.dtype == dtype`?
Using short-circuit evaluation is better.